services/sso/src/main/java/software.amazon.awssdk.services.sso/internal/SsoCredentialsUtils.java

```diff
@@ -0,0 +1,165 @@
+import software.amazon.awssdk.profiles.Profile;
+import software.amazon.awssdk.services.sso.SsoClient;
+import software.amazon.awssdk.services.sso.auth.AwsSsoCredentials;
+import software.amazon.awssdk.services.sso.model.GetRoleCredentialsRequest;
+import software.amazon.awssdk.services.sso.model.GetRoleCredentialsResponse;
+import software.amazon.awssdk.services.sso.model.RoleCredentials;
+import software.amazon.awssdk.utils.Validate;
+import software.amazon.awssdk.utils.internal.SystemSettingUtils;
+
+@SdkProtectedApi
+public final class SsoCredentialsUtils {
+
+    private static final String TOKEN_DIRECTORY = "/Users/" + SystemSettingUtils.resolveSetting(USER_NAME).get() +
+                                                  "/.aws/sso/cache/";
+
+    private Profile profile;
+    private SsoClient ssoClient;
+
+    private String ssoRegion;
+    private String ssoAccountId;
+    private String ssoRoleName;
+    private String ssoStartUrl;
+
+    private SsoCredentialsTokenProperties tokenProperties;
+
+    public SsoCredentialsUtils(Profile profile, SsoClient ssoClient) {
+        this.profile = profile;
+        this.ssoClient = ssoClient;
+    }
+
+    /**
+     * Validate and resolve the SSO profile. Then load the cached token file to retrieve necessary information for the SSO call.
+     */
+    public AwsCredentials getUpdatedCredentials() {
+        validateAndResolveSsoProfile(profile);
+
+        try {
+            String cachedTokenPath = generateCachedTokenPath(ssoStartUrl);
+            File cachedTokenFile = new File(cachedTokenPath);
+
+            tokenProperties = new ObjectMapper().readValue(cachedTokenFile,
+                                                           SsoCredentialsTokenProperties.class);
+            if (tokenExpired(tokenProperties.getExpiresAt())) {
+                throw ExpiredTokenException.builder().message("The SSO session associated with this profile has expired or is"
+                                                              + " otherwise invalid. To refresh this SSO session run aws sso"
+                                                              + " login with the corresponding profile.").build();
+            }
+
+        } catch (IOException e) {
+            throw SdkClientException.builder().message("Unable to read the cached token file. Please check if the correct "
+                                                       + "sso_start_url is configured in the profile").cause(e).build();
+        }
+
+        return getCredentialsFromSso();
+    }
+
+    private void validateAndResolveSsoProfile(Profile profile) {
+        Map<String, String> configurations = profile.properties();
+
+        if (configurations.containsKey(SSO_ACCOUNT_ID)) {
+            ssoAccountId = configurations.get(SSO_ACCOUNT_ID);
+        } else {
+            throw SdkClientException.create(String.format("This profile is invalid, %s is missing.", SSO_ACCOUNT_ID));
+        }
```

Comment:
This is a bit DRYer, and makes it only one hash key lookup in `configurations` instead of two:

```Java
ssoAccountId = Validate.notNull(configurations.get(SSO_ACCOUNT_ID), 
                                "The %s profile is invalid, %s is missing.", profile.name(), SSO_ACCOUNT_ID);
```